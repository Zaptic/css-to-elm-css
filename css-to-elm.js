const postcss = require('postcss');
const scss = require('postcss-scss');
const fs = require('fs');

function outputNode(node) {
  if (node.type === 'rule') {
    const text = `
${node.selector} : Style
${node.selector} =
    Css.batch [
`;

    console.log(text);

    for (let child of node.nodes) {
      outputNode(child);
    }

    console.log(`    ]`);
  } else if (node.type === 'decl') {
    if (node.prop === 'border') {
      const values = node.value.split(' ');
      console.log(`    border${values.length} ${node.value}`);
    }
  }
}

fs.readFile('./View.cssm.scss', (err, css) => {
  postcss([])
    .process(css, {syntax: scss})
    .then(result => {
      for (let node of result.root.nodes) {
        outputNode(node);
        // console.log(JSON.stringify(node, null, 2));
      }
    });
});
