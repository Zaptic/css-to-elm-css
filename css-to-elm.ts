import postcss from 'postcss'
import scss from 'postcss-scss'
import * as fs from 'fs'
import slug from 'slug'
import * as _ from 'lodash'

import * as Rules from './rules'

interface ElmNode {
    name: string
    decls: Rules.ElmDecl[]
}

function convertDecl(node: postcss.Declaration): Rules.ElmDecl {
    if (node.type === 'decl') {
        const handler = Rules.lookup[node.prop] || Rules.standard
        return handler(node)
    }
    return { name: '', values: [] }
}

function convertNode(node: postcss.Node): ElmNode[] {
    if (node.type === 'rule') {
        const children: ElmNode[] = []
        const decls = []
        for (let child of node.nodes || []) {
            if (child.type === 'rule') {
                const childNodes = convertNode(child).map(childElmNode => ({
                    ...childElmNode,
                    name: `${node.selector}-${childElmNode.name}`,
                }))
                children.push(...childNodes)
            } else if (child.type === 'decl') {
                decls.push(convertDecl(child))
            }
        }
        return [{ name: node.selector, decls }, ...children]
    }
    return []
}

function elmNodeToString(node: ElmNode): string {
    const rules = node.decls
        .map((decl: Rules.ElmDecl) => {
            return `${decl.name} ${decl.values.join(' ')}`
        })
        .join('\n        , ')

    // Remove greater-than & ampersand syntax from the concatenated name
    const simplified = node.name.replace(/(&|>)/g, '')

    // Slugify to remove anything else & then camelCase for Elm
    const name = _.camelCase(slug(simplified))

    const text = `
${name} : Style
${name} =
    Css.batch
        [ ${rules}
        ]
`
    return text
}

const cssFilePath = process.argv[2]

fs.readFile(cssFilePath, (err, css) => {
    postcss([])
        .process(css, { syntax: scss })
        .then((result: any) => {
            for (let node of result.root.nodes) {
                const elmNodes = convertNode(node)
                elmNodes.map(elmNode => console.log(elmNodeToString(elmNode)))
            }
        })
})
