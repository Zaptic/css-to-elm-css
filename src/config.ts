export interface Config {
    replacements: { [key: string]: string }
    imports: string[]
    unnest: boolean
}

export const defaultConfig: Config = {
    /* Replacements allow direct substitutions on values. Helpful for handling SCSS variables and replacing them with
       your own Elm variables, eg:

           replacements: {
               '$zaptic-grey-5': 'Color.grey5',
               '$zaptic-grey-7': 'Color.grey7',
               '$zaptic-red': 'Color.red',
           },

    */
    replacements: {},

    /* Imports allow you to specify addition imports that you'd like included in the output. These might be used to
       expose standard colors that you have defined in your SCSS:

           imports: [
               'import Zap.Style.Color as Color'
           ],
    */
    imports: [],

    unnest: false,
}
