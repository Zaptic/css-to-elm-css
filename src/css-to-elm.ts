import * as fs from 'fs'
import * as path from 'path'

import * as _ from 'lodash'
import postcss from 'postcss'
import scss from 'postcss-scss'
import slug from 'slug'
import program from 'commander'

import * as Rules from './rules'
import { Config, defaultConfig } from './config'

interface ElmRule {
    type: 'rule'
    name: string
    children: ElmNode[]
}

interface ElmComment {
    type: 'comment'
    content: string[]
}

type ElmNode = ElmRule | ElmComment | Rules.ElmDecl

function extractSource(node: postcss.Node) {
    if (node.source.start && node.source.end) {
        const lines = (<any>node.source.input).css.split('\n')
        return lines.slice(node.source.start.line - 1, node.source.end.line)
    }
    return []
}

function convertDecl(node: postcss.Declaration, config: Config): Rules.ElmDecl {
    if (node.type === 'decl') {
        const handler = Rules.lookup[node.prop] || Rules.standard
        return handler(node, config)
    }
    return { type: 'decl', name: '', values: [] }
}

function convertNode(node: postcss.Node, config: Config): ElmNode[] {
    if (node.type === 'rule') {
        const children: ElmNode[] = []
        for (let child of node.nodes || []) {
            if (child.type === 'rule') {
                const childNodes = convertNode(child, config).map(childElmNode => {
                    if (childElmNode.type === 'rule' && config.unnest) {
                        return { ...childElmNode, name: `${node.selector}-${childElmNode.name}` }
                    } else {
                        return childElmNode
                    }
                })
                children.push(...childNodes)
            } else if (child.type === 'decl') {
                children.push(convertDecl(child, config))
            } else {
                children.push({ type: 'comment', content: extractSource(child) })
            }
        }

        if (config.unnest) {
            return [{ type: 'rule', name: node.selector, children: [] }, ...children]
        } else {
            return [{ type: 'rule', name: node.selector, children }]
        }
    } else if (node.type === 'decl') {
        // Assume it is a variable declaration so strip the leading '$' and camelCase the rest
        return [
            {
                type: 'decl',
                name: _.camelCase(node.prop.replace('$', '')),
                values: [Rules.translateValue(node.value, config)],
            },
        ]
    } else {
        return [{ type: 'comment', content: extractSource(node) }]
    }
    return []
}

/* We want to turn something like 

      > button,
      a

   into ['> button', 'a']. In particular, it breaks on 'commas' so that we can each selector separately.
*/
function extractSelectors(name: string) {
    return name
        .replace('\n', '')
        .replace(/ +/g, ' ')
        .split(',')
        .map(str => str.trim())
}

const symbolLookup: { [key: string]: string } = {
    '>': 'Global.children',
}

const globalRegex = /^:global\((.*)\)/

/* Converts a selector like '> button' into:

   [ 'Global.children', 'Global.typeSelector "button"']

 */
function convertSelector(selector: string) {
    // Try to handle inserting 'decescendants' as needed by checking that there is no leading '&'
    // There are probably lots of ways of improving this
    const lead = selector[0] === '&' ? [] : ['Global.descendants <| List.singleton <|']

    // Test for entry string main entirely of lower case letters or h1 to h6
    const tagRegex = /^([a-z]+|h[1-6])$/

    return [
        ...lead,
        ...selector.split(' ').map(entry => {
            const globalMatch = entry.match(globalRegex)
            if (entry[0] === '.') {
                return `Global.class "${_.camelCase(entry)}"`
            } else if (symbolLookup[entry] !== undefined) {
                return symbolLookup[entry]
            } else if (globalMatch) {
                return `Global.class "${globalMatch[1].slice(1)}"`
            } else if (tagRegex.test(entry)) {
                return `Global.typeSelector "${entry}"`
            } else {
                return `Global.selector "${entry}"`
            }
        }),
    ]
}

function elmRuleContentsToString(node: ElmRule): string {
    const rules = node.children
        .map((child: ElmNode, index: number) => {
            if (child.type === 'decl') {
                const sep = index ? ',' : '['
                return `${sep} ${child.name} ${child.values.join(' ')}`
            } else if (child.type === 'rule') {
                const names = extractSelectors(child.name)
                const all = names.map(convertSelector)

                const contents = elmRuleContentsToString(child)
                return [
                    ...all.map((selector, selectorIndex) => {
                        const sep = index + selectorIndex ? ', ' : '['
                        return [sep, selector.join(' '), contents].join('')
                    }),
                ].join(' ')
            } else if (child.type === 'comment') {
                const sep = index ? ', ' : '['
                return `${sep} unknownNode "${child.content}"`
            }
        })
        .join('\n        ')

    const text = `
        ${rules || '['}
        ]
`
    return text
}

function elmNodeToString(node: ElmNode): string {
    if (node.type === 'rule') {
        // Remove greater-than & ampersand syntax from the concatenated name
        const simplified = node.name.replace(/(&|>)/g, '')

        // Slugify to remove anything else & then camelCase for Elm
        const name = _.camelCase(slug(simplified))

        const text = `
${name} : Style
${name} =
    Css.batch`
        return text + elmRuleContentsToString(node)
    } else if (node.type === 'decl') {
        return `${node.name} = ${node.values.join(' ')}`
    } else if (node.type === 'comment') {
        return node.content.map(line => '-- ' + line).join('\n')
    }
    return ''
}

function header(config: Config) {
    return `
module Style exposing (..)

import Css exposing (..)')
import Css.Global as Global')
${config.imports.join('\n')}
`
}

function helpers() {
    return `

unknownNode string = Css.batch []

merge ops styles =
    let
        reduce entry prev =
            entry [ prev ]
    in
    List.foldr reduce (Css.batch styles) ops
`
}

export function convertCss(css: string, config: Config) {
    return postcss([])
        .process(css, { syntax: scss })
        .then((result: any) => {
            const content = result.root.nodes.map((node: postcss.Node) => {
                return convertNode(node, config).map(elmNodeToString)
            })

            return content.join('\n')
        })
}

async function main() {
    program
        .version('0.1.0')
        .option('-c, --config [path]', 'Path to configuration file')
        .parse(process.argv)

    const cssFilePath = program.args[0]
    const configFilePath = program.config
    let config = defaultConfig

    if (configFilePath !== undefined) {
        config = {
            ...config,
            ...require(path.resolve(configFilePath)),
        }
    }

    fs.readFile(cssFilePath, async (err, css) => {
        const output = await convertCss(css.toString(), config)
        console.log(header(config))
        console.log(helpers())
        console.log(output)
    })
}

if (process.env.NODE_ENV !== 'test') {
    main()
}
