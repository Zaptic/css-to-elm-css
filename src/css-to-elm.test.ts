import { convertCss } from './css-to-elm'
import { defaultConfig } from './config'

test('converts an empty string', async () => {
    expect(await convertCss('', defaultConfig)).toBe('')
})

test('basic class selector', async () => {
    const input = `
.myButton {
    color: white;
}
    `
    expect(await convertCss(input, defaultConfig)).toMatchSnapshot()
})

test('nested class selectors', async () => {
    const input = `
.myButton {
    color: white;

    .myButtonDetail {
        color: black;
    }
}
    `
    expect(await convertCss(input, defaultConfig)).toMatchSnapshot()
})

test('readme content nested', async () => {
    const input = `
.trainingWorkflows {
    padding: 30px 20px;
    text-align: center;

    width: $workflow-column-width;

    background-color: #f8fbff;

    border-right: 1px solid $zaptic-grey-6;

    h2 {
        font-size: 16px;
        padding: 25px 20px 20px 20px;
        margin: 0;
        border-bottom: 1px solid $zaptic-grey-5;
    }

    > ul {
        padding: 0;
        margin: 0;
        width: 100%;
    }
}
    `
    expect(await convertCss(input, { ...defaultConfig, unnest: false })).toMatchSnapshot()
})

test('readme content unnested', async () => {
    const input = `
.trainingWorkflows {
    padding: 30px 20px;
    text-align: center;

    width: $workflow-column-width;

    background-color: #f8fbff;

    border-right: 1px solid $zaptic-grey-6;

    h2 {
        font-size: 16px;
        padding: 25px 20px 20px 20px;
        margin: 0;
        border-bottom: 1px solid $zaptic-grey-5;
    }

    > ul {
        padding: 0;
        margin: 0;
        width: 100%;
    }
}
    `
    expect(await convertCss(input, { ...defaultConfig, unnest: true })).toMatchSnapshot()
})
