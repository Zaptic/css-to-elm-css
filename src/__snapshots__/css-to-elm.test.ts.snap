// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic class selector 1`] = `
"
myButton : Style
myButton =
    Css.batch
        [ color white
        ]
"
`;

exports[`nested class selectors 1`] = `
"
myButton : Style
myButton =
    Css.batch
        [ color white
        , Global.descendants <| List.singleton <| Global.class \\"myButtonDetail\\" 
        [ color black
        ]

        ]
"
`;

exports[`readme content nested 1`] = `
"
trainingWorkflows : Style
trainingWorkflows =
    Css.batch
        [ padding2 (px 30) (px 20)
        , textAlign center
        , width workflowColumnWidth
        , backgroundColor (hex \\"f8fbff\\")
        , borderRight3 (px 1) solid zapticGrey6
        , Global.descendants <| List.singleton <| Global.typeSelector \\"h2\\"
        [ fontSize (px 16)
        , padding4 (px 25) (px 20) (px 20) (px 20)
        , margin zero
        , borderBottom3 (px 1) solid zapticGrey5
        ]

        , Global.descendants <| List.singleton <| Global.children Global.typeSelector \\"ul\\"
        [ padding zero
        , margin zero
        , width (pct 100)
        ]

        ]
"
`;

exports[`readme content unnested 1`] = `
"
trainingWorkflows : Style
trainingWorkflows =
    Css.batch
        [
        ]
,padding2 = (px 30) (px 20),textAlign = center,width = workflowColumnWidth,backgroundColor = (hex \\"f8fbff\\"),borderRight3 = (px 1) solid zapticGrey6,
trainingWorkflowsH2 : Style
trainingWorkflowsH2 =
    Css.batch
        [
        ]
,fontSize = (px 16),padding4 = (px 25) (px 20) (px 20) (px 20),margin = zero,borderBottom3 = (px 1) solid zapticGrey5,
trainingWorkflowsUl : Style
trainingWorkflowsUl =
    Css.batch
        [
        ]
,padding = zero,margin = zero,width = (pct 100)"
`;
